### PR TITLE
Keys to parts

### PR DESCRIPTION
**Pull request by manushT**

### CODE REVIEW

File: internal/core/input.rs [hunk 1]
```diff
@@ -664,6 +683,7 @@ fn keys_from_parts_inner<'a>(
                 }
                 ignore_alt = true;
             }
+            "Physical" => is_physical = true,
```
Comment (LeonMatthes) [inline internal/core/input.rs]: Nit: I prefer if this is `@physical-keys` to match the macro name.

File: internal/core/input.rs [hunk 2]
```diff
@@ -675,13 +695,36 @@ fn keys_from_parts_inner<'a>(
...
+            .ok_or_else(|| KeysParseErrorInner::UnknownPhysicalKey(key_name.into()))?;
+        return Ok(Keys {
+            inner: KeysInner {
+                key: native.into(),
+                modifiers,
+                ignore_shift,
+                ignore_alt,
+                is_physical: true,
+            },
+        });
```
Comment (LeonMatthes) [inline internal/core/input.rs]: nit: I prefer to move this branch, as well as the following code into separate functions so that the control flow is clearer, e.g.:

```slint
if is_physical {
    Self::from_physical_parts(key_name, modifiers, ignore_shift, ignore_alt)
} else {
    Self::from_logical_parts(key_name, modifiers, ignore_shift, ignore_alt)
}
```

File: internal/core/input.rs [hunk 3]
```diff
@@ -752,6 +795,11 @@ impl Keys {
     /// if not found, treated as a string literal (must be a single lowercase grapheme cluster).
     /// Exactly one non-modifier key must be present.
     ///
+    /// To create a physical-key shortcut (the runtime counterpart of
+    /// `@physical-keys(...)`), include the marker `"Physical"` in the parts list. The
+    /// key part is then resolved against the physical-key table (`A`, `Digit1`,
+    /// `BackQuote`, `LeftArrow`, ...) and string-literal fallback is not allowed.
```
Comment (LeonMatthes) [inline internal/core/input.rs]: Let's tighten the contract here and enforce the first entry of the parts list must be `"Physical"` (or `@physical-keys`).

That would disambiguate if we ever do allow string literals in physical keys, as then you could even have a key named "@physical-keys" with:
`["@physical-keys", "@physical-keys"]`

It's probably overly cautious, but I think it's better to be cautious than to allow stuff that shouldn't be allowed.

File: internal/core/input.rs [hunk 4]
```diff
@@ -763,6 +811,88 @@ impl Keys {
...
+            }
+            parts.push("Physical".into());
+            // Reverse-map the native physical-key name (e.g. "Backquote") back to
+            // the user-facing spelling accepted by `from_parts` ("BackQuote").
+            parts.push(match lookup_physical_key_native(inner.key.as_str()) {
+                Some(name) => name.into(),
+                None => inner.key.clone(),
+            });
+            return parts;
+        }
```
Comment (LeonMatthes) [inline internal/core/input.rs]: Again, let's please split the program flow more clearly here.

Use two helper functions with a clear branch:
```rust
if inner.is_physical {
    self.to_physical_parts(parts)
} else {
    self.to_logical_parts(parts)
}
```
(or something similar, feel free to structure the helper functions differently).

File: internal/core/input.rs [hunk 5]
```diff
@@ -763,6 +811,88 @@ impl Keys {
...
+            parts.push("Alt?".into());
+        }
+
+        if inner.is_physical {
+            if inner.modifiers.shift {
+                parts.push("Shift".into());
+            } else if inner.ignore_shift {
+                parts.push("Shift?".into());
+            }
+            parts.push("Physical".into());
```
Comment (LeonMatthes) [inline internal/core/input.rs]: This (or `@physical-keys` should be appended to the front, so we mark the behavior from the start).

File: internal/core/input.rs [hunk 6]
```diff
@@ -675,13 +695,36 @@ fn keys_from_parts_inner<'a>(
...
         None => return Err(KeysParseErrorInner::NoKey),
     };
 
+    if is_physical {
+        // @physical-keys(...) syntax: resolve the key against the physical-key table;
+        // string literals are not allowed. Accept either the user-facing spelling
+        // (e.g. "BackQuote") or the native spelling (e.g. "Backquote") so that the
+        // native fallback emitted by `to_parts` round-trips.
+        let native = lookup_physical_key_name(key_name)
+            .or_else(|| lookup_physical_key_native(key_name).map(|_| key_name))
```
Comment (LeonMatthes) [inline internal/core/input.rs]: We should only accept one or the other.
This is meant to be used programmatically, not for "pretty printing".

I need to check why we have the two different names in the first place, that actually seems a bit odd 🤔
(That's for the initial PR, not this one though).

File: internal/core/input.rs [hunk 7]
```diff
@@ -2989,6 +3119,87 @@ mod tests {
...
+            assert_eq!(k, k2, "round-trip mismatch for {parts:?} → {out_strs:?}");
+        }
+    }
+
+    #[test]
+    fn test_physical_native_spelling_roundtrip() {
+        // `from_parts` accepts the native physical spelling (e.g. "Backquote")
+        // in addition to the user-facing one ("BackQuote"); both resolve to the
+        // same physical `Keys`. This guarantees the native fallback that
+        // `to_parts` can emit re-parses into an equal value.
```
Comment (LeonMatthes) [inline internal/core/input.rs]: Again, please only accept one or the other.

File: internal/core/input.rs [hunk 8]
```diff
@@ -2989,6 +3119,87 @@ mod tests {
...
+        let cases: &[(&[&str], &[&str])] = &[
+            (&[], &[]),
+            (&["A"], &["A"]),
+            (&["a"], &["A"]), // letter literal → named form
+            (&["Control", "S"], &["Control", "S"]),
+            (&["Control", "Shift?", "Z"], &["Control", "Shift?", "Z"]),
+            (&["Control", "Alt?", "A"], &["Control", "Alt?", "A"]),
+            (&["F5"], &["F5"]),
+            (&["Control", "Plus"], &["Control", "Plus"]), // no Shift? for LocalizedShiftable
+            (&["Control", "+"], &["Control", "Plus"]),    // literal '+' → named "Plus"
```
Comment (LeonMatthes) [inline internal/core/input.rs]: Hm, this does not roundtrip cleanly, does it?

As `["Control", "+"]` does *not* have ignore_shift set, while `["Control", "Plus"]` does.

Please add this to the roundtrip test to confirm this indeed does not roundtrip correctly.

The simple fix is to not do any reverse-lookup on the logical key codes.
So this:
```
Keys::from_parts(["Control", "Plus"]).unwrap().into_parts();
```
would "desugar" to: `["Control", "Shift?", "+"]`, which is totally fine, and now cleanly roundtrips.

We just have to make sure that we specify the contract correctly:
`Key -> parts -> Key` roundtrips cleanly.
`parts -> Keys -> parts` does *not* roundtrip, as the initial from_parts may apply de-sugaring, which we cannot undo anymore.

That also means we can remove the reverse-lookup code again.

**Review by LeonMatthes — CHANGES_REQUESTED** [review]
I belive this has a bug, as it tries to undo the "desugaring" we apply when parsing, which is unnecessary and I believe is currently incorrect.

`from_parts(["Control", "Pus"])` should turn into these parts: `["Control", "Shift?", "+"]`.

File: internal/core/input.rs (hunk 1 above)
Comment (manushT) [reply internal/core/input.rs]: done

File: internal/core/input.rs (hunk 2 above)
Comment (manushT) [reply internal/core/input.rs]: done

File: internal/core/input.rs (hunk 5 above)
Comment (manushT) [reply internal/core/input.rs]: done

File: internal/core/input.rs (hunk 4 above)
Comment (manushT) [reply internal/core/input.rs]: done

File: internal/core/input.rs (hunk 7 above)
Comment (manushT) [reply internal/core/input.rs]: done

File: internal/core/input.rs (hunk 8 above)
Comment (manushT) [reply internal/core/input.rs]: done

Comment (manushT) [reply internal/core/input.rs]: partly

Comment (manushT) [reply internal/core/input.rs]: Why not remove *all* reverse-lookup (the literal suggestion): `Unshiftable`
  whitespace keys store chars that `from_parts` trims away — Space `' '`,
  Tab `'\t'`, Return `'\n'` — and function keys store private-use codepoints.
  Emitting those raw would break the `Keys → parts → Keys` round-trip and
  produce unreadable output, so the name is retained for `Unshiftable` keys.

File: internal/core/input.rs [hunk 9]
```diff
@@ -610,25 +633,40 @@ fn normalize_key(key: &str) -> SharedString {
...
+            // The marker only declares the kind of shortcut, and only as the very
+            // first entry, so the parts list is unambiguous from the start.
+            if seen_part {
+                return Err(KeysParseErrorInner::PhysicalMarkerNotFirst);
+            }
+            is_physical = true;
+            seen_part = true;
+            continue;
+        }
+        seen_part = true;
```
Comment (LeonMatthes) [inline internal/core/input.rs]: Instead of setting a flag whether we have "seen a part", please just use `enumerate` on the iterator and assert that we're on the index 0.
Currently this causes a bug where `[" ", "@physical-keys", "Space"]` would be accepted, even if `@physical-keys` is not the first part.

File: internal/core/input.rs [hunk 10]
```diff
@@ -763,6 +851,106 @@ impl Keys {
...
+            parts.push("Shift".into());
+        } else if inner.ignore_shift {
+            parts.push("Shift?".into());
+        }
+    }
+
+    fn to_physical_parts(&self) -> alloc::vec::Vec<SharedString> {
+        let mut parts = alloc::vec::Vec::new();
+        // The marker leads the list, mirroring `from_parts`' requirement.
+        parts.push(PHYSICAL_KEYS_MARKER.into());
```
Comment (LeonMatthes) [inline internal/core/input.rs]:
```suggestion
        let mut parts = vec![PHYSICAL_KEYS_MARKER.to_shared_string()];
```

File: internal/core/input.rs (hunk 8 above)
Comment (LeonMatthes) [reply internal/core/input.rs]: Ah, I would lean to remove the whitespace trimming then.

`Keys::from_parts([" "])` sounds like it should be a valid way to produce a space logical key IMO. `@keys(" ")` is also valid in Slint.
Or do you see any problems with this? :thinking: 

(I know it's technically a breaking change, but I don't expect this to cause problems TBH).

**Review by LeonMatthes — COMMENTED** [review]
Looking much better :)
Still some nitpicks to do.

The biggest remaining question is whether we should remove the whitespace trimming so we can safely desugar.

File: internal/core/input.rs (hunk 8 above)
Comment (manushT) [reply internal/core/input.rs]: removed the reverse-lookup and whitespace trimming

File: examples/runtime_key_bindings/cpp/main.cpp [hunk 11]
```diff
@@ -3,14 +3,198 @@
...
 #include <iostream>
+#include <iterator>
+#include <optional>
+#include <sstream>
+#include <string>
 #include <string_view>
 #include <vector>
 
+namespace {
+
```
Comment (LeonMatthes) [inline examples/runtime_key_bindings/cpp/main.cpp]: This example is starting to grow quite a bit.
Most of it is duplicated with the Rust example.

For us, the only relevant part of the example is demonstrating we can indeed call `toParts` from C++.
The rest is just demo C++ code that we need to maintain.

I prefer removing the C++ part of the example and just sticking with Rust.
It still shows how to conceptually load/store user shortcuts.
There is not much value in doing this again in C++.

If you want to persist the example, you can link to the sha1 or this PR in the README.

File: examples/runtime_key_bindings/rust/main.rs [hunk 12]
```diff
@@ -7,16 +7,132 @@
...
+    let mut out = String::with_capacity(part.len());
+    for c in part.chars() {
+        if c.is_ascii_graphic() && c != '%' {
+            out.push(c);
+        } else {
+            let mut buf = [0u8; 4];
+            for b in c.encode_utf8(&mut buf).as_bytes() {
+                out.push('%');
+                out.push(HEX[(b >> 4) as usize] as char);
+                out.push(HEX[(b & 0xf) as usize] as char);
```
Comment (LeonMatthes) [inline examples/runtime_key_bindings/rust/main.rs]: nit: this can be easier done with:

```suggestion
                out.push(format!("%{b:02X}"));
```

File: internal/core/input.rs [hunk 13]
```diff
@@ -759,6 +778,75 @@ impl Keys {
...
+    /// A `Keys` value round-trips through `from_parts`:
+    ///
+    /// ```
+    /// use i_slint_core::input::Keys;
+    /// let k = Keys::from_parts(["Control", "Shift?", "Z"])?;
+    /// let parts = k.to_parts();
+    /// // The key comes back as the stored character, not as the `Z` name it was built from.
+    /// assert_eq!(parts.iter().map(|s| s.as_str()).collect::<Vec<_>>(),
+    ///            vec!["Control", "Shift?", "z"]);
+    /// assert_eq!(Keys::from_parts(parts.iter().map(|s| s.as_str()))?, k);
```
Comment (LeonMatthes) [inline internal/core/input.rs]: This over-specifies, please remove it.

The only thing we should guarantee is:
```suggestion
    /// let k_from_parts = Key::from_parts(parts);
    /// assert_eq!(k_from_parts, k);
```

We intentionally want to keep the representation inside Keys private.

The only thing we should specify is that `Keys` values round-trip like this:
> A `Keys` value that is converted into parts and then re-created by those parts with `from_parts` will be equal to the input `Keys` value.

We intentionally do *not* guarantee that the parts themselves are equal.

File: internal/core/input.rs [hunk 14]
```diff
@@ -759,6 +778,75 @@ impl Keys {
...
+    ///
+    /// The key is always emitted as the character the `Keys` stores, never as a key
+    /// name: `["A"]`, `["Space"]` and `["F5"]` come back as `["a"]`, `[" "]` and the
+    /// F5 character. Modifiers keep their names, since that is how they are stored.
+    ///
+    /// Note the round-trip direction: `Keys → to_parts → from_parts` reproduces an
+    /// equal `Keys`. The reverse (`parts → from_parts → to_parts`) is *not*
+    /// guaranteed to reproduce the input parts, because `from_parts` de-sugars names
+    /// (e.g. `["Control", "Plus"]` becomes `["Control", "Shift?", "+"]`), and
+    /// de-sugaring is one-way.
```
Comment (LeonMatthes) [inline internal/core/input.rs]: This also overspecifies, but the core message is right.

Maybe specify it instead as:

> Note that while a round-trip guarantees that the resulting `Keys` instances will be equal, the parts returned by `to_parts` can be different from the parts used to construct the `Keys` instance with `from_parts`.

File: internal/core/input.rs [hunk 15]
```diff
@@ -759,6 +778,75 @@ impl Keys {
...
+    ///
+    /// Because a part may be a space, a control character or a private-use codepoint,
+    /// a text format storing these needs to quote or escape them; see the
+    /// `runtime_key_bindings` example.
```
Comment (LeonMatthes) [inline internal/core/input.rs]: Please add a link to the example, otherwise it is not discoverable from docs.rs.

File: internal/core/input.rs [hunk 16]
```diff
@@ -2955,6 +3043,36 @@ mod tests {
                 false,
             ),
             ("A alone (named key)", &["A"], "a", KeyboardModifiers::default(), false, false),
+            // Whitespace is significant: these literals name the Space/Tab/Return keys
+            // and must agree with their named spellings (parts are not trimmed).
+            ("\" \" literal → Space", &[" "], " ", KeyboardModifiers::default(), false, false),
+            ("Space named", &["Space"], " ", KeyboardModifiers::default(), false, false),
+            ("\"\\t\" literal → Tab", &["\t"], "\t", KeyboardModifiers::default(), false, false),
```
Comment (LeonMatthes) [inline internal/core/input.rs:3072]: Also add a test for one of the special unicode reserved characters we use to represent e.g. F5/Play/Pause, etc.

**Review by LeonMatthes — COMMENTED** [review]
Thank you, we should be able to merge this soon.

The description is promising a bit too much, and I think the C++ example is growing to more code than it is worth, given we have the same example in Rust.

So once those are addressed we should be able to merge.

File: examples/runtime_key_bindings/cpp/main.cpp (hunk 11 above)
Comment (manushT) [reply examples/runtime_key_bindings/cpp/main.cpp]: all good, lets drop it :)

File: examples/runtime_key_bindings/rust/main.rs (hunk 12 above)
Comment (manushT) [reply examples/runtime_key_bindings/rust/main.rs]: yep nice :)

File: internal/core/input.rs (hunk 13 above)
Comment (manushT) [reply internal/core/input.rs]: done

File: internal/core/input.rs (hunk 14 above)
Comment (manushT) [reply internal/core/input.rs]: done

File: internal/core/input.rs (hunk 15 above)
Comment (manushT) [reply internal/core/input.rs]: done

File: internal/core/input.rs (hunk 16 above)
Comment (manushT) [reply internal/core/input.rs:3072]: done

File: internal/core/input.rs [hunk 17]
```diff
@@ -759,6 +778,69 @@ impl Keys {
...
+    /// Decompose this `Keys` value into the list of string parts that
+    /// [`Keys::from_parts`] accepts.
+    ///
+    /// See also the Slint documentation on [Key Bindings](slint:KeyBindingOverview).
+    ///
+    /// A `Keys` value that is converted into parts and then re-created from those parts
+    /// with [`Keys::from_parts`] will be equal to the input `Keys` value:
+    ///
+    /// ```
+    /// use i_slint_core::input::Keys;
```
Comment (tronical) [inline internal/core/input.rs]: We should not use `i_slint_core` in the public Rust API.

Comment (ogoffart) [reply internal/core/input.rs]: Something like that 

```suggestion
    /// # mod slint { use i_slint_core::input::Keys; }
    /// use slint::Keys;
```

Comment (ogoffart) [reply internal/core/input.rs]:
```suggestion
    /// use slint::Keys;
```

should also work because slint is a dev dependency

**Review by LeonMatthes — APPROVED** [review]
